File: code/leetcode/leetcode_39/leetcode_39.cpp
class Solution {
public:
    void backtrace(vector<int> & candidates, int target, int start_index, vector<int> &mid, vector<vector<int>> &res){
        if(target == 0){
            res.push_back(mid);
            return;
        }
        for(int i = start_index; i < candidates.size(); i++){
            if(target >=  candidates[i]){
                mid.push_back(candidates[i]);
                backtrace(candidates, target - candidates[i], i, mid, res);
                mid.pop_back();
            }else{
                return;
            }
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res = {};
        vector<int> mid = {};
        backtrace(candidates, target, 0, mid, res);
        return res;
    }
};

File: code/leetcode/leetcode_39/leetcode_39.rs
impl Solution {
    pub fn backtrace(candidates: &Vec<i32>, target: i32, start_idx: usize, mid:  &mut Vec<i32>, res:  & mut Vec<Vec<i32>>){
        if target == 0{
            res.push(mid.to_vec());
            return;
        }

        for i in start_idx..candidates.len(){
            if target >= candidates[i]{
                mid.push(candidates[i]);
                Self::backtrace(candidates, target - candidates[i], i, mid, res);
                mid.pop();
            }else{
                return;
            }
        }
    }
    pub fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        candidates.sort();
        let mut res =  Vec::new();
        Self::backtrace(&candidates, target, 0, &mut vec![], &mut res);
        return res;
    }
}